src/cdomains/intDomain.ml

```diff
@@ -1887,15 +2444,198 @@ module IntDomTupleImpl = struct
+      , opt_map2 (r.f2p (module I2)) xb yb
+      , opt_map2 (r.f2p (module I3)) xc yc
+      , opt_map2 (r.f2p (module I4)) xd yd )
+
+  (* f2p: binary projections *)
+  let (%%) f g x = f % (g x) (* composition for binary function g *)
+
+  let leq =
+    for_all
+    %% map2p {f2p= (fun (type a) (module I : S with type t = a) ?no_ov -> I.leq)}
+
+  let equal =
+    for_all
+    %% map2p {f2p= (fun (type a) (module I : S with type t = a) ?no_ov -> I.equal)}
+
+  let compare =
+    List.fold_left (fun a x -> if x <> 0 then x else a) 0
+    % to_list
+    %% map2p {f2p= (fun (type a) (module I : S with type t = a) ?no_ov -> I.compare)} (* idea? same impl. as above... *)
+
+
+  let flat f x = match to_list_some x with [] -> None | xs -> Some (f xs)
+
+  let to_excl_list x = mapp2 { fp2 = fun (type a) (module I:S with type t = a and type int_t = int_t) -> I.to_excl_list } x |> flat List.concat
+
+  let to_incl_list x =
+    let hd l = match l with h::t -> h | _ -> [] in
+    let tl l = match l with h::t -> t | _ -> [] in
+    let a y = BatSet.of_list (hd y) in
+    let b y = BatList.map BatSet.of_list (tl y) in
+    let merge y = BatSet.elements @@ BatList.fold BatSet.intersect (a y) (b y)
+    in
+    mapp2 { fp2 = fun (type a) (module I:S with type t = a and type int_t = int_t) -> I.to_incl_list } x |> flat merge
+
+
+  let pretty () = (fun xs -> text "(" ++ (try List.reduce (fun a b -> a ++ text "," ++ b) xs with _ -> nil) ++ text ")") % to_list % mapp { fp = fun (type a) (module I:S with type t = a) -> (* assert sf==I.short; *) I.pretty () } (* NOTE: the version above does something else. also, we ignore the sf-argument here. *)
+
+
+  let refine_functions ik : (t -> t) list =
+    let maybe reffun ik domtup dom =
+      match dom with Some y -> reffun ik domtup y | _ -> domtup
+    in
+    [(fun (a, b, c, d) -> refine_with_excl_list ik (a, b, c, d) (to_excl_list (a, b, c, d)));
+     (fun (a, b, c, d) -> refine_with_incl_list ik (a, b, c, d) (to_incl_list (a, b, c, d)));
+     (fun (a, b, c, d) -> maybe refine_with_interval ik (a, b, c, d) b);
+     (fun (a, b, c, d) -> maybe refine_with_congruence ik (a, b, c, d) d)]
+
+  let refine ik ((a, b, c, d ) : t ) : t =
+    let dt = ref (a, b, c, d) in
+    let _ = match GobConfig.get_string "ana.int.refinement" with
+      | "never" -> ()
+      | "once" ->
+         List.iter (fun f -> dt := f !dt) (refine_functions ik);
+      | "fixpoint" ->
+         let quit_loop = ref false in
+         while not !quit_loop do
+           let old_dt = !dt in
+           List.iter (fun f -> dt := f !dt) (refine_functions ik);
+           quit_loop := equal old_dt !dt;
+           if is_bot !dt then dt := bot_of ik; quit_loop := true;
+           if M.tracing then M.trace "cong-refine-loop" "old: %a, new: %a\n" pretty old_dt pretty !dt;
+         done;
+      | _ -> ()
+    in
+    !dt
+
+  let no_overflow ik r =
+    if GobConfig.get_bool "ana.int.congruence_no_overflow" then true
```

Comment:
`ana.int.congruence_no_overflow` should only apply to **signed** overflows, overflows fro variables of unsigned integer types are well-defined and may exist in programs without Undefined Behavior.

Reply:
I think the reason why I blankly applied it to all integers is that we lack a coherent theory of how the congruence classes change when overflowing.
For example (0 + 5ℤ) +(1 + 5ℤ) = (1 + 5ℤ). But, assuming we're working with 4 bit to represent unsigned integers, 15 + 1 = 0 ∈ (0 + 5ℤ). So unless we're working with singletons, I think it's still necessary to know if an overflow might occur.

Reply:
>  So unless we're working with singletons, I think it's still necessary to know if an overflow might occur.

Yes, I absolutely agree that we need to know if an overflow can occur. However, currently if `ana.int.congruence_no_overflow` is enabled, it just claims that there are NEVER any overflows. This is justified for signed types (as overflows there are undefined behavior, and one may want to assume that programs are free of that).

However, unsigned types have well-defined overflows, so one can not just claim that they never overflow. Instead for unsigned types - even if I assume no UB - I need to use the logic to check for overflows via Intervals that is also used when `ana.int.congruence_no_overflow` is disabled.

Or am I misunderstanding your point?

Reply:
Ah, then I misunderstood the semantics the flag should have. I thought of it as ignoring any and all integer over- or underflows.

Of course if this should only ignore undefined overflows, this completely makes sense.